### PR TITLE
Test hidden `nonce` content attribute behavior.

### PR DESCRIPTION
**Pull request by mikewest**
As specified in whatwg/html#2373.

### CODE REVIEW

**Comment by mikewest** [comment]
First pass at tests for https://github.com/whatwg/html/pull/2373. What else would y'all like to see, @annevk, @arturjanc, and @zcorpan?

**Comment by annevk** [comment]
This all looks resaonable to me. Wouldn't getComputedStyle be an easier way to test CSS? Or did you specifically want to test that event? It seems with that event you don't know if it failed if the browser decided not to fetch the image for some other reason.

**Comment by w3c-bots** [comment]
[View the complete job log.](https://travis-ci.org/w3c/web-platform-tests/jobs/219645465)

# Firefox (nightly channel) #
Testing web-platform-tests at revision 0525a5e7d1e98750b6ed750562c82fbd49babc13
Using browser at version BuildID 20170406100208; SourceStamp 3c68d659c2b715f811708f043a1e7169d77be2ba
Starting 10 test iterations
All results were stable
## All results ##
<details>
<summary>1 test ran</summary>
<details>
<summary><a href="http://w3c-test.org/submissions/5423/content-security-policy/script-src/script-nonces-hidden.html">/content-security-policy/script-src/script-nonces-hidden.html</a></summary>


|                           Subtest                            | Results |                                                      Messages                                                     |
|--------------------------------------------------------------|---------|-------------------------------------------------------------------------------------------------------------------|
|                                                              | OK      |                                                                                                                   |
| `HTML: Reading 'nonce' content attribute and IDL attribute.` | FAIL    | `assert_equals: expected Element node <script nonce="abc" id="testScript" executed="yay">\n  doc... but got null` |
| `HTML: Cloned node retains nonce.`                           | FAIL    | `assert_equals: IDL attribute expected (string) "abc" but got (undefined) undefined`                              |
| `HTML: Cloned node retains nonce when inserted.`             | FAIL    | `assert_equals: expected (string) "abc" but got (undefined) undefined`                                            |
| `HTML: Writing 'nonce' content attribute.`                   | FAIL    | `assert_equals: expected (string) "abc" but got (undefined) undefined`                                            |
| `HTML: Writing 'nonce' IDL attribute.`                       | PASS    |                                                                                                                   |
| `Document-written script executes.`                          | PASS    |                                                                                                                   |
| `HTML: Document-written script's nonce value.`               | FAIL    | `assert_equals: expected "" but got "abc"`                                                                        |
| `HTML: createElement.nonce.`                                 | FAIL    | `assert_equals: expected (string) "yay" but got (object) null`                                                    |
| `HTML: createElement.setAttribute.`                          | FAIL    | `assert_equals: Pre-insertion IDL expected (string) "" but got (undefined) undefined`                             |
| `Nonces don't leak via CSS side-channels.`                   | PASS    |                                                                                                                   |

</details>
</details>

**Comment by w3c-bots** [comment]
[View the complete job log.](https://travis-ci.org/w3c/web-platform-tests/jobs/219645468)

# Chrome (unstable channel) #
Testing web-platform-tests at revision 0525a5e7d1e98750b6ed750562c82fbd49babc13
Using browser at version 59.0.3063.4 dev
Starting 10 test iterations
All results were stable
## All results ##
<details>
<summary>1 test ran</summary>
<details>
<summary><a href="http://w3c-test.org/submissions/5423/content-security-policy/script-src/script-nonces-hidden.html">/content-security-policy/script-src/script-nonces-hidden.html</a></summary>


|                           Subtest                            | Results |                                                      Messages                                                     |
|--------------------------------------------------------------|---------|-------------------------------------------------------------------------------------------------------------------|
|                                                              | OK      |                                                                                                                   |
| `HTML: Reading 'nonce' content attribute and IDL attribute.` | FAIL    | `assert_equals: expected Element node <script nonce="abc" id="testScript" executed="yay">\n  doc... but got null` |
| `HTML: Cloned node retains nonce.`                           | FAIL    | `assert_equals: IDL attribute expected (string) "abc" but got (undefined) undefined`                              |
| `HTML: Cloned node retains nonce when inserted.`             | FAIL    | `assert_equals: expected (string) "abc" but got (undefined) undefined`                                            |
| `HTML: Writing 'nonce' content attribute.`                   | FAIL    | `assert_equals: expected (string) "abc" but got (undefined) undefined`                                            |
| `HTML: Writing 'nonce' IDL attribute.`                       | PASS    |                                                                                                                   |
| `Document-written script executes.`                          | PASS    |                                                                                                                   |
| `HTML: Document-written script's nonce value.`               | FAIL    | `assert_equals: expected "" but got "abc"`                                                                        |
| `HTML: createElement.nonce.`                                 | FAIL    | `assert_equals: expected (string) "yay" but got (object) null`                                                    |
| `HTML: createElement.setAttribute.`                          | FAIL    | `assert_equals: Pre-insertion IDL expected (string) "" but got (undefined) undefined`                             |
| `Nonces don't leak via CSS side-channels.`                   | PASS    |                                                                                                                   |

</details>
</details>

**Comment by mikewest** [comment]
Ah, I didn't think about `getComputedStyle`. :) That's simpler.

**Comment by w3c-bots** [comment]
These tests are now available on [w3c-test.org](https://w3c-test.org/submissions/5423/)

**Comment by mikewest** [comment]
Updated the last test to use `getComputedStyle`. Not sure if the `requestAnimationFrame` is necessary; I have no idea what the timing of CSS style resolution/application is in relation to the current JavaScript task...

**Comment by annevk** [comment]
getComputedStyle forces style resolution, but the loading of background images is not deterministic, so that event might never fire I suppose.

**Comment by annevk** [comment]
(Does it fire on time in browsers today?)

**Comment by mikewest** [comment]
It doesn't. :( I'll drop the event. It's just going to be flaky even if it works, and the `getComputedStyle` bit is good enough in itself.

File: content-security-policy/script-src/script-nonces-hidden.html [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+</script>
+
+<!-- CSS Leakage -->
+<style>
+    #cssTest { display: block; }
+    #cssTest[nonce=abc] { background: url(/security/resources/abe.png); }
+</style>
+<script nonce="abc" id="cssTest">
+    async_test(t => {
+      requestAnimationFrame(t.step_func_done(_ => {
```
Comment (zcorpan) [inline content-security-policy/script-src/script-nonces-hidden.html]: No need for rAF. Can just make this a sync `test()`.

File: content-security-policy/script-src/script-nonces-hidden.html [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+      var s = document.createElement('script');
+      s.innerText = script.innerText;
+      s.setAttribute('nonce', 'abc');
+      assert_equals(s.getAttribute('nonce'), 'abc', "Pre-insertion content");
+      assert_equals(s.nonce, '', "Pre-insertion IDL");
+      document.head.appendChild(s);
+
+      window.addEventListener('load', t.step_func_done(_ => {
+        assert_equals(s.nonce, 'abc', "Post-insertion IDL");
+        assert_equals(s.getAttribute('nonce'), '', "Post-insertion content");
```
Comment (zcorpan) [inline content-security-policy/script-src/script-nonces-hidden.html]: I think the above two should be asserted directly after `document.head.appendChild(s);`

File: content-security-policy/script-src/script-nonces-hidden.html [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+    // Create node.
+    async_test(t => {
+      var s = document.createElement('script');
+      s.innerText = script.innerText;
+      s.nonce = 'abc';
+      document.head.appendChild(s);
+
+      window.addEventListener('load', t.step_func_done(_ => {
+        assert_equals(s.nonce, 'abc');
+        assert_equals(s.getAttribute('nonce'), null);
```
Comment (zcorpan) [inline content-security-policy/script-src/script-nonces-hidden.html]: I think the above two should be asserted directly after document.head.appendChild(s);

File: content-security-policy/script-src/script-nonces-hidden.html [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+        assert_equals(script.getAttribute('nonce'), '');
+        assert_equals(script.nonce, 'abc');
+      }, "HTML: Document-written script's nonce value.");
+    </scr` + `ipt>`);
+
+    // Create node.
+    async_test(t => {
+      var s = document.createElement('script');
+      s.innerText = script.innerText;
+      s.nonce = 'abc';
```
Comment (zcorpan) [inline content-security-policy/script-src/script-nonces-hidden.html]: Assert pre-insertion IDL and content attribute here?

**Comment by annevk** [comment]
This is now #21853. I didn't address @zcorpan's nits though and they still seem valid...

**Comment by annevk** [comment]
Remaining nits now addressed in the other PR. Some were already addressed.

**Pull request closed without merging**